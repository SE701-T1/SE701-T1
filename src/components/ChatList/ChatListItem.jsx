import React from 'react';
import { ButtonBase, Typography } from '@mui/material';
import PropTypes from 'prop-types';
import ProfileBadge from '../ProfileBadge/ProfileBadge';
import styles from './ChatListItem.module.css';
import timeSince from '../../Util/TimeFormater';

/**
 * ChatItemList is the component that display each of the messages in the chat.
 */
function ChatItemList({ selected, name, active, lastMessageText, lastMessageTime, onClick }) {
  const onContainerClick = () => {
    if (onClick !== undefined) {
      onClick();
    }
  };

  return (
    <ButtonBase
      data-testid="ButtonBase"
      style={{ background: selected ? '#ECF3FF' : '#FFFFFF', padding: '5px', borderRadius: '5px' }}
      className={styles.container}
      onClick={() => onContainerClick()}>
      <div className={styles.profileBadgeContainer}>
        <ProfileBadge active={active} name={name} />
      </div>
      <div className={styles.textContainer}>
        <div>
          <Typography
            sx={{
              fontSize: '20px',
              fontStyle: 'normal',
              fontWeight: '600',
              lineHeight: '28px',
            }}
            className={styles.headerText}>
            {name}
          </Typography>
        </div>
        <div className={styles.subText}>
          <Typography
            className={styles.lastMessageText}
            color="#959595"
            variant="subtitle2"
            lineHeight="normal">
            {lastMessageText}
          </Typography>
          <Typography color="#959595" variant="subtitle2" lineHeight="normal">
            â€¢ {timeSince(lastMessageTime)}
          </Typography>
        </div>
      </div>
    </ButtonBase>
  );
}

ChatItemList.propTypes = {
  name: PropTypes.string.isRequired,
  active: PropTypes.bool.isRequired,
  lastMessageText: PropTypes.string.isRequired,
  lastMessageTime: PropTypes.instanceOf(Date).isRequired,
  onClick: PropTypes.func.isRequired,
};

export default ChatItemList;
