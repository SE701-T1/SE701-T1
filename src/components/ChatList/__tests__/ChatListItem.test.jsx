import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ChatListItem from '../ChatListItem';

describe('ChatListItem', () => {
  const activeUserData = {
    id: 1,
    name: 'name_1',
    lastMessage: 'last_message_1',
    lastMessageTime: new Date(),
    active: true,
  };

  it('is correctly rendered to the screen', () => {
    render(
      <ChatListItem
        name={activeUserData.name}
        active={activeUserData.active}
        lastMessageText={activeUserData.lastMessage}
        lastMessageTime={activeUserData.lastMessageTime}
        onClick={() => null}
      />,
    );

    expect(screen.getByText('name_1')).toBeInTheDocument();
    expect(screen.getByText('last_message_1')).toBeInTheDocument();
  });

  it('Click chat list item changes it to selected', () => {
    render(
      <ChatListItem
        testId="under_test"
        name={activeUserData.name}
        active={activeUserData.active}
        lastMessageText={activeUserData.lastMessage}
        lastMessageTime={activeUserData.lastMessageTime}
        onClick={() => null}
      />,
    );

    fireEvent.click(screen.getByTestId('ButtonBase'));

    expect(screen.getByTestId('ButtonBase')).toHaveStyle({ background: 'rgb(255, 255, 255)' });
  });

  it('produces the correct initial snapshot', () => {
    const tree = renderer
      .create(
        <ChatListItem
          testId="under_test"
          name={activeUserData.name}
          active={activeUserData.active}
          lastMessageText={activeUserData.lastMessage}
          lastMessageTime={activeUserData.lastMessageTime}
          onClick={() => null}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
