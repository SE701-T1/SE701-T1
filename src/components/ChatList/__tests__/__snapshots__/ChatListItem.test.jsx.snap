// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatListItem produces the correct initial snapshot 1`] = `
<button
  className="MuiButtonBase-root container css-10d1a0h-MuiButtonBase-root"
  data-testid="ButtonBase"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "background": "#FFFFFF",
      "borderRadius": "5px",
      "padding": "5px",
    }
  }
  tabIndex={0}
  type="button"
>
  <div
    className="profileBadgeContainer"
  >
    <div
      className="MuiContainer-root MuiContainer-maxWidthLg MuiContainer-disableGutters css-6ch1n3-MuiContainer-root"
    >
      <span
        className="MuiBadge-root css-1srwgqi-MuiBadge-root"
        data-testid="active"
      >
        <div
          className="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1pl8iwh-MuiAvatar-root"
          data-testid="name_1"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
        <span
          className="MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight css-ank3r8-MuiBadge-badge"
        >
          
        </span>
      </span>
    </div>
  </div>
  <div
    className="textContainer"
  >
    <div>
      <p
        className="MuiTypography-root MuiTypography-body1 headerText css-1x6dds0-MuiTypography-root"
      >
        name_1
      </p>
    </div>
    <div
      className="subText"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 lastMessageText css-1i1udgb-MuiTypography-root"
      >
        last_message_1
      </h6>
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 css-1i1udgb-MuiTypography-root"
      >
        â€¢ 
        0 s
      </h6>
    </div>
  </div>
</button>
`;
